cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/edge/Dataflow.kt

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2024, Fraunhofer AISEC. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *                    $$$$$$\  $$$$$$$\   $$$$$$\
+ *                   $$  __$$\ $$  __$$\ $$  __$$\
+ *                   $$ /  \__|$$ |  $$ |$$ /  \__|
+ *                   $$ |      $$$$$$$  |$$ |$$$$\
+ *                   $$ |      $$  ____/ $$ |\_$$ |
+ *                   $$ |  $$\ $$ |      $$ |  $$ |
+ *                   \$$$$$   |$$ |      \$$$$$   |
+ *                    \______/ \__|       \______/
+ *
+ */
+package de.fraunhofer.aisec.cpg.graph.edge
+
+import de.fraunhofer.aisec.cpg.graph.Node
+import de.fraunhofer.aisec.cpg.graph.declarations.FieldDeclaration
+import org.neo4j.ogm.annotation.RelationshipEntity
+
+/**
+ * The granularity of the data-flow, e.g., whether the flow contains the whole object, or just a
+ * part of it, for example a record (class/struct) member. In the latter case, the part can be
+ * specified using the [Dataflow.memberField], which contains the field declaration node.
+ */
+enum class GranularityType {
+    FULL,
+    PARTIAL
+}
+
+/**
+ * This edge class defines a flow of data between [start] and [end]. The flow can have a certain
+ * [granularity].
+ */
+@RelationshipEntity
+class Dataflow(
+    start: Node,
+    end: Node,
+    val granularity: GranularityType = GranularityType.FULL,
+    val memberField: FieldDeclaration? = null,
```

Comment:
Could we also directly add the `index` and `calling_context_in` and `calling_context_out` features? Do we really want this `legacyProperties` map? I'd prefer the ability to generate the map from the class' properties in a getter or make the data available to the common functions of the `PropertyEdge` base-class if necessary for backwards compatibility.

Reply:
No we don't want this legacy property map, at least not in the "public" functions I can have a look, whether I can generate the map behind the scenes. I will probably not get rid of it all-to-gether at this point.